data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/AvroConsumer.java

```diff
@@ -0,0 +1,103 @@
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.kafka.configuration.KafkaSourceConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.TopicsConfig;
+import org.opensearch.dataprepper.plugins.kafka.source.KafkaSourceBufferAccumulator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * * A helper class which helps to process the avro records and write them into the buffer.
+ * Offset handling and consumer re-balance are also being handled
+ */
+
+@SuppressWarnings("deprecation")
+public class AvroConsumer implements KafkaSourceSchemaConsumer<String, GenericRecord>, ConsumerRebalanceListener {
+    private static final Logger LOG = LoggerFactory.getLogger(AvroConsumer.class);
+    private Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = new HashMap<>();
+    private long lastReadOffset = 0L;
+    private KafkaConsumer<String, GenericRecord> kafkaAvroConsumer;
+    private volatile long lastCommitTime = System.currentTimeMillis();
+
+    @Override
+    public void consumeRecords(final KafkaConsumer<String, GenericRecord> consumer, final AtomicBoolean status,
+                               final Buffer<Record<Object>> buffer, final TopicsConfig topicConfig,
+                               final KafkaSourceConfig kafkaSourceConfig, PluginMetrics pluginMetrics,
+                               final String schemaType) {
+        KafkaSourceBufferAccumulator kafkaSourceBufferAccumulator = new KafkaSourceBufferAccumulator(topicConfig,kafkaSourceConfig,
+                schemaType,pluginMetrics);
+        kafkaAvroConsumer = consumer;
+        try {
+            consumer.subscribe(Arrays.asList(topicConfig.getName()));
+            while (!status.get()) {
+                offsetsToCommit.clear();
+                ConsumerRecords<String, GenericRecord> records = consumer.poll(Duration.ofMillis(100));
+                if (!records.isEmpty() && records.count() > 0) {
+                    for (TopicPartition partition : records.partitions()) {
+                        List<Record<Object>> kafkaRecords = new ArrayList<>();
+                        List<ConsumerRecord<String, GenericRecord>> partitionRecords = records.records(partition);
+                        for (ConsumerRecord<String, GenericRecord> consumerRecord : partitionRecords) {
+                            offsetsToCommit.put(new TopicPartition(consumerRecord.topic(), consumerRecord.partition()),
+                                    new OffsetAndMetadata(consumerRecord.offset() + 1, null));
+                            kafkaRecords.add(kafkaSourceBufferAccumulator.getEventRecord(consumerRecord.value().toString()));
```

Comment:
Is toString() on GenericRecord reliable way to convert data into JSON object? Should we use JsonEncoder instead? There are some known issues in toString() https://issues.apache.org/jira/browse/AVRO-3061

Reply:
I think we can't use json encoder because we need to create json parser which takes string as parameter.